internal/storage/stats.go

```diff
@@ -29,12 +33,78 @@ const (
 	maxBloomFalsePositive float64 = 0.005
 )
 
-// Int64Stats contains statistics data for int64 column
-type Int64Stats struct {
+type PrimaryKey struct {
+	id *schemapb.ID
+}
+
+func (p *PrimaryKey) GetID() *schemapb.ID {
+	return p.id
+}
+
+func (p *PrimaryKey) MarshalJSON() ([]byte, error) {
+	m := jsonpb.Marshaler{}
+	var buffer bytes.Buffer
+	err := m.Marshal(io.Writer(&buffer), p.id)
+	if err != nil {
+		return nil, err
+	}
+
+	return buffer.Bytes(), nil
+}
+
+func (p *PrimaryKey) UnmarshalJSON(data []byte) error {
+	id := &schemapb.ID{}
+	err := jsonpb.Unmarshal(bytes.NewReader(data), id)
+	if err != nil {
+		return err
+	}
+
+	p.id = id
+	return nil
+}
+
+// PrimaryKeyStats contains statistics data for pk column
+type PrimaryKeyStats struct {
 	FieldID int64              `json:"fieldID"`
-	Max     int64              `json:"max"`
-	Min     int64              `json:"min"`
+	Max     int64              `json:"max"` // useless, will delete
+	Min     int64              `json:"min"` //useless, will delete
 	BF      *bloom.BloomFilter `json:"bf"`
+	MaxPk   *PrimaryKey        `json:"maxPk"`
+	MinPk   *PrimaryKey        `json:"minPk"`
+}
+
+func (stats *PrimaryKeyStats) updatePk(pk *PrimaryKey) {
+	if stats.MinPk == nil {
+		stats.MinPk = pk
+	}
+
+	if stats.MaxPk == nil {
+		stats.MaxPk = pk
+	}
+
+	minID := stats.MinPk.GetID()
+	maxID := stats.MaxPk.GetID()
+	pkID := pk.GetID()
+
+	switch pkID.GetPrimaryKey().(type) {
+	case *schemapb.ID_LongPk:
+		if pkID.GetLongPk() < minID.GetLongPk() {
+			stats.MinPk = pk
+		}
+		if maxID.GetLongPk() < pkID.GetLongPk() {
+			stats.MaxPk = pk
+		}
+
+	case *schemapb.ID_StringPk:
+		if pkID.GetStringPk() < minID.GetStringPk() {
```

Comment:
the string PK compare function might be wrong for ""

Reply:
use strings pkg to compare(strings.Compare(str1, str2)